{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import roc_auc_score, roc_curve, average_precision_score, precision_recall_curve\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_17369/1154448997.py:1: DtypeWarning: Columns (3,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  features = pd.read_table('/work3/s220260/PhosKing1.0/data/clean_data/features.tsv').rename(columns={'#UniProt-ID': 'id', 'position': 'pos'})\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>residue</th>\n",
       "      <th>kinases</th>\n",
       "      <th>species</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8RXG3</td>\n",
       "      <td>12</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EPSD,PhosPhAt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q8RXG3</td>\n",
       "      <td>6</td>\n",
       "      <td>S</td>\n",
       "      <td>ASK7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q8RXG3</td>\n",
       "      <td>215</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UniProt,EPSD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q8RXG3</td>\n",
       "      <td>221</td>\n",
       "      <td>S</td>\n",
       "      <td>ASK7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UniProt,EPSD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q8RXG3</td>\n",
       "      <td>225</td>\n",
       "      <td>T</td>\n",
       "      <td>ASK7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UniProt,EPSD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645019</th>\n",
       "      <td>Z4YM84</td>\n",
       "      <td>371</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EPSD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645020</th>\n",
       "      <td>Z4YM84</td>\n",
       "      <td>4</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EPSD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645021</th>\n",
       "      <td>Z4YMB8</td>\n",
       "      <td>768</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EPSD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645022</th>\n",
       "      <td>Z4YN06</td>\n",
       "      <td>4</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EPSD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645023</th>\n",
       "      <td>Z4YN62</td>\n",
       "      <td>166</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EPSD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1645024 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             id  pos residue kinases species        sources\n",
       "0        Q8RXG3   12       S     NaN     NaN  EPSD,PhosPhAt\n",
       "1        Q8RXG3    6       S    ASK7     NaN        UniProt\n",
       "2        Q8RXG3  215       T     NaN     NaN   UniProt,EPSD\n",
       "3        Q8RXG3  221       S    ASK7     NaN   UniProt,EPSD\n",
       "4        Q8RXG3  225       T    ASK7     NaN   UniProt,EPSD\n",
       "...         ...  ...     ...     ...     ...            ...\n",
       "1645019  Z4YM84  371       S     NaN     NaN           EPSD\n",
       "1645020  Z4YM84    4       T     NaN     NaN           EPSD\n",
       "1645021  Z4YMB8  768       S     NaN     NaN           EPSD\n",
       "1645022  Z4YN06    4       T     NaN     NaN           EPSD\n",
       "1645023  Z4YN62  166       S     NaN     NaN           EPSD\n",
       "\n",
       "[1645024 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features = pd.read_table('/work3/s220260/PhosKing1.0/data/clean_data/features.tsv').rename(columns={'#UniProt-ID': 'id', 'position': 'pos'})\n",
    "features_ids = features['id'].unique()\n",
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0          Q8RXG3\n",
       " 7          Q8L793\n",
       " 26         Q2V3V8\n",
       " 28         Q8GWB3\n",
       " 31         Q9C505\n",
       "             ...  \n",
       " 1645016    Z4YLR9\n",
       " 1645017    Z4YM84\n",
       " 1645021    Z4YMB8\n",
       " 1645022    Z4YN06\n",
       " 1645023    Z4YN62\n",
       " Name: id, Length: 229088, dtype: object]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_seq_org = features[['id', \"species\"]].drop_duplicates()\n",
    "features_seq_org = [features_seq_org['id']]\n",
    "features_seq_org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylogeny</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D4A7L7</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P70365</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P55096</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K2P5</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P61968</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205776</th>\n",
       "      <td>P53567</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205777</th>\n",
       "      <td>Q8VG33</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205778</th>\n",
       "      <td>Q9LUL2</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Eukaryota Viridiplantae Streptophyta Embryophy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205779</th>\n",
       "      <td>Q9BQ48</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205780</th>\n",
       "      <td>P33048</td>\n",
       "      <td>Capra hircus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205781 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id              organism  \\\n",
       "0           D4A7L7     Rattus norvegicus   \n",
       "1           P70365          Mus musculus   \n",
       "2           P55096          Mus musculus   \n",
       "3       A0A0G2K2P5     Rattus norvegicus   \n",
       "4           P61968          Homo sapiens   \n",
       "...            ...                   ...   \n",
       "205776      P53567          Homo sapiens   \n",
       "205777      Q8VG33          Mus musculus   \n",
       "205778      Q9LUL2  Arabidopsis thaliana   \n",
       "205779      Q9BQ48          Homo sapiens   \n",
       "205780      P33048          Capra hircus   \n",
       "\n",
       "                                                phylogeny  \n",
       "0       Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "1       Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "2       Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "3       Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "4       Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "...                                                   ...  \n",
       "205776  Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "205777  Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "205778  Eukaryota Viridiplantae Streptophyta Embryophy...  \n",
       "205779  Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "205780  Eukaryota Metazoa Chordata Craniata Vertebrata...  \n",
       "\n",
       "[205781 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_organisms = pd.read_csv('/work3/s220260/PhosKing1.0/data/clean_data/seq_organisms.csv', header=None).rename(columns={0: 'id', 1: 'organism', 2: 'phylogeny'}).drop_duplicates()\n",
    "seq_organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_organism(row: pd.Series):\n",
    "    organism = row['organism']\n",
    "    phylogeny = row['phylogeny']\n",
    "    conditions = {\n",
    "        'human': 'homo' in organism.lower() or 'homo' in phylogeny.lower(),\n",
    "        'mammalian': 'mammalia' in phylogeny.lower() or 'mus musculus' in organism.lower() or 'rattus' in organism.lower() or 'sus scrofa' in organism.lower(),\n",
    "        'other_animal': 'metazoa' in phylogeny.lower(),\n",
    "        'bacteria': 'bacteria' in phylogeny.lower(),\n",
    "        'fungal': 'fungi' in phylogeny.lower(),\n",
    "        'plant': 'plantae' in phylogeny.lower(),\n",
    "        'other_eukaryotes': 'amoebozoa' in phylogeny.lower() or 'sar' in phylogeny.lower() or 'discoba' in phylogeny.lower() or 'cryptophyceae' in phylogeny.lower() or 'rhodophyta' in phylogeny.lower() or 'haptista' in phylogeny.lower() or 'giardia' in phylogeny.lower() or 'cyanophora' in phylogeny.lower(),\n",
    "        'virus': 'virus' in organism.lower() or 'virus' in phylogeny.lower(),\n",
    "        'archaea': 'archaea' in phylogeny.lower(),\n",
    "    }\n",
    "    priority = ['human', 'mammalian', 'other_animal', 'bacteria', 'fungal', 'plant', 'other_eukaryotes', 'virus', 'archaea']\n",
    "    for org_group in priority:\n",
    "        if conditions[org_group]:\n",
    "            return org_group\n",
    "    return None\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylogeny</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D4A7L7</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P70365</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P55096</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K2P5</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P61968</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205776</th>\n",
       "      <td>P53567</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205777</th>\n",
       "      <td>Q8VG33</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205778</th>\n",
       "      <td>Q9LUL2</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Eukaryota Viridiplantae Streptophyta Embryophy...</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205779</th>\n",
       "      <td>Q9BQ48</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205780</th>\n",
       "      <td>P33048</td>\n",
       "      <td>Capra hircus</td>\n",
       "      <td>Eukaryota Metazoa Chordata Craniata Vertebrata...</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205781 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id              organism  \\\n",
       "0           D4A7L7     Rattus norvegicus   \n",
       "1           P70365          Mus musculus   \n",
       "2           P55096          Mus musculus   \n",
       "3       A0A0G2K2P5     Rattus norvegicus   \n",
       "4           P61968          Homo sapiens   \n",
       "...            ...                   ...   \n",
       "205776      P53567          Homo sapiens   \n",
       "205777      Q8VG33          Mus musculus   \n",
       "205778      Q9LUL2  Arabidopsis thaliana   \n",
       "205779      Q9BQ48          Homo sapiens   \n",
       "205780      P33048          Capra hircus   \n",
       "\n",
       "                                                phylogeny      group  \n",
       "0       Eukaryota Metazoa Chordata Craniata Vertebrata...  mammalian  \n",
       "1       Eukaryota Metazoa Chordata Craniata Vertebrata...  mammalian  \n",
       "2       Eukaryota Metazoa Chordata Craniata Vertebrata...  mammalian  \n",
       "3       Eukaryota Metazoa Chordata Craniata Vertebrata...  mammalian  \n",
       "4       Eukaryota Metazoa Chordata Craniata Vertebrata...      human  \n",
       "...                                                   ...        ...  \n",
       "205776  Eukaryota Metazoa Chordata Craniata Vertebrata...      human  \n",
       "205777  Eukaryota Metazoa Chordata Craniata Vertebrata...  mammalian  \n",
       "205778  Eukaryota Viridiplantae Streptophyta Embryophy...      plant  \n",
       "205779  Eukaryota Metazoa Chordata Craniata Vertebrata...      human  \n",
       "205780  Eukaryota Metazoa Chordata Craniata Vertebrata...  mammalian  \n",
       "\n",
       "[205781 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_organisms['group'] = seq_organisms.apply(group_organism, axis=1)\n",
    "seq_organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unclassified organisms:\n",
      "unclassified sequences metagenomes ecological metagenomes\n",
      "unclassified sequences environmental samples\n",
      "unclassified sequences metagenomes\n",
      "\n",
      "Unclassified phylogenies:\n",
      "bioreactor metagenome\n",
      "mine drainage metagenome\n",
      "uncultured organism\n",
      "uncultured organism Bio4\n",
      "metagenome\n"
     ]
    }
   ],
   "source": [
    "missing_group = seq_organisms[seq_organisms['group'].isna()]\n",
    "print('Unclassified organisms:')\n",
    "print(*missing_group['phylogeny'].unique().tolist(), sep='\\n')\n",
    "print('\\nUnclassified phylogenies:')\n",
    "print(*missing_group['organism'].unique().tolist(), sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "  \n",
    "true_phosphos = defaultdict(set)\n",
    "true_phosphos.update(features.groupby('id')['pos'].apply(set))\n",
    "\n",
    "def get_truth(row: pd.Series):\n",
    "    return int(row['pos'] in true_phosphos[row['id']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUC: 0.865\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.472</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>5</td>\n",
       "      <td>0.551</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>9</td>\n",
       "      <td>0.474</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>17</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>24</td>\n",
       "      <td>0.395</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205410</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>195</td>\n",
       "      <td>0.232</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205411</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>196</td>\n",
       "      <td>0.247</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205412</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>204</td>\n",
       "      <td>0.128</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205413</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>206</td>\n",
       "      <td>0.105</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205414</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>209</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205415 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id  pos  score  y_true\n",
       "0       Q5PPN7    2  0.472       0\n",
       "1       Q5PPN7    5  0.551       0\n",
       "2       Q5PPN7    9  0.474       0\n",
       "3       Q5PPN7   17  0.433       0\n",
       "4       Q5PPN7   24  0.395       0\n",
       "...        ...  ...    ...     ...\n",
       "205410  Q58FZ9  195  0.232       0\n",
       "205411  Q58FZ9  196  0.247       0\n",
       "205412  Q58FZ9  204  0.128       0\n",
       "205413  Q58FZ9  206  0.105       0\n",
       "205414  Q58FZ9  209  0.079       0\n",
       "\n",
       "[205415 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "netphos4_test = pd.read_table('pre_model1_test_.tsv', header=None).rename(columns={0: 'id', 1: 'pos', 2: '?', 3: 'score'}).drop('?', axis=1)\n",
    "assert len(netphos4_test) == len(netphos4_test[netphos4_test['id'].isin(features_ids)]), 'Mismatch between features and test set'\n",
    "netphos4_test['y_true'] = netphos4_test.apply(get_truth, axis=1)\n",
    "assert len(netphos4_test[netphos4_test['y_true'].isna()]) == 0\n",
    "print(f'AUC: {roc_auc_score(netphos4_test[\"y_true\"], netphos4_test[\"score\"]):.3f}')\n",
    "netphos4_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUC: 0.759\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P73120</td>\n",
       "      <td>12</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P73120</td>\n",
       "      <td>14</td>\n",
       "      <td>0.530</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P73120</td>\n",
       "      <td>20</td>\n",
       "      <td>0.264</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P73120</td>\n",
       "      <td>25</td>\n",
       "      <td>0.138</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P73120</td>\n",
       "      <td>28</td>\n",
       "      <td>0.194</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98456</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>353</td>\n",
       "      <td>0.503</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98457</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>355</td>\n",
       "      <td>0.654</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98458</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>377</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98459</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>387</td>\n",
       "      <td>0.667</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98460</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>395</td>\n",
       "      <td>0.704</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98461 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  pos  score  y_true\n",
       "0      P73120   12  0.240       0\n",
       "1      P73120   14  0.530       0\n",
       "2      P73120   20  0.264       0\n",
       "3      P73120   25  0.138       0\n",
       "4      P73120   28  0.194       0\n",
       "...       ...  ...    ...     ...\n",
       "98456  P0DPA2  353  0.503       0\n",
       "98457  P0DPA2  355  0.654       0\n",
       "98458  P0DPA2  377  0.064       0\n",
       "98459  P0DPA2  387  0.667       1\n",
       "98460  P0DPA2  395  0.704       0\n",
       "\n",
       "[98461 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "musitedeep_test = pd.read_table('Test_MusiteDeep_1000.tsv', header=None).rename(columns={0: 'id', 1: 'pos', 2: '?', 3: 'score'}).drop('?', axis=1)\n",
    "assert len(musitedeep_test) == len(musitedeep_test[musitedeep_test['id'].isin(features_ids)]), 'Mismatch between features and test set'\n",
    "musitedeep_test['y_true'] = musitedeep_test.apply(get_truth, axis=1)\n",
    "assert len(musitedeep_test[musitedeep_test['y_true'].isna()]) == 0\n",
    "print(f'AUC: {roc_auc_score(musitedeep_test[\"y_true\"], musitedeep_test[\"score\"]):.3f}')\n",
    "musitedeep_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUC: 0.626\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A088AGW7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.440</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A088AGW7</td>\n",
       "      <td>4</td>\n",
       "      <td>0.512</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A088AGW7</td>\n",
       "      <td>13</td>\n",
       "      <td>0.443</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A088AGW7</td>\n",
       "      <td>22</td>\n",
       "      <td>0.516</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A088AGW7</td>\n",
       "      <td>26</td>\n",
       "      <td>0.443</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107464</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2732</td>\n",
       "      <td>0.816</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107465</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2743</td>\n",
       "      <td>0.660</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107466</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2753</td>\n",
       "      <td>0.544</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107467</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2759</td>\n",
       "      <td>0.439</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107468</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2772</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107469 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id   pos  score  y_true\n",
       "0       A0A088AGW7     3  0.440       0\n",
       "1       A0A088AGW7     4  0.512       0\n",
       "2       A0A088AGW7    13  0.443       0\n",
       "3       A0A088AGW7    22  0.516       0\n",
       "4       A0A088AGW7    26  0.443       0\n",
       "...            ...   ...    ...     ...\n",
       "107464      V4Z445  2732  0.816       0\n",
       "107465      V4Z445  2743  0.660       0\n",
       "107466      V4Z445  2753  0.544       0\n",
       "107467      V4Z445  2759  0.439       0\n",
       "107468      V4Z445  2772  0.403       0\n",
       "\n",
       "[107469 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "netphos3_test = pd.read_table('Test_NetPhos3.1_1000.tsv', header=None).rename(columns={0: 'id', 1: 'pos', 2: '?', 3: 'score'}).drop('?', axis=1)\n",
    "assert len(netphos3_test) == len(netphos3_test[netphos3_test['id'].isin(features_ids)]), 'Mismatch between features and test set'\n",
    "netphos3_test['y_true'] = netphos3_test.apply(get_truth, axis=1)\n",
    "assert len(netphos3_test[netphos3_test['y_true'].isna()]) == 0\n",
    "print(f'AUC: {roc_auc_score(netphos3_test[\"y_true\"], netphos3_test[\"score\"]):.3f}')\n",
    "netphos3_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUC: 0.740\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>W4XZI0</td>\n",
       "      <td>13</td>\n",
       "      <td>0.772</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>W4XZI0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>W4XZI0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.121</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W4XZI0</td>\n",
       "      <td>27</td>\n",
       "      <td>0.257</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>W4XZI0</td>\n",
       "      <td>35</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35454</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>464</td>\n",
       "      <td>0.224</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35455</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>465</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35456</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>467</td>\n",
       "      <td>0.172</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35457</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>469</td>\n",
       "      <td>0.097</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35458</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>475</td>\n",
       "      <td>0.137</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35459 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  pos  score  y_true\n",
       "0      W4XZI0   13  0.772       1\n",
       "1      W4XZI0   21  0.124       0\n",
       "2      W4XZI0   24  0.121       0\n",
       "3      W4XZI0   27  0.257       0\n",
       "4      W4XZI0   35  0.317       0\n",
       "...       ...  ...    ...     ...\n",
       "35454  Q9GU50  464  0.224       0\n",
       "35455  Q9GU50  465  0.203       0\n",
       "35456  Q9GU50  467  0.172       0\n",
       "35457  Q9GU50  469  0.097       0\n",
       "35458  Q9GU50  475  0.137       0\n",
       "\n",
       "[35459 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "netphospan_test = pd.read_table('Test_NetPhospan1.0.tsv', header=None).rename(columns={0: 'id', 1: 'pos', 2: '?', 3: 'score'}).drop('?', axis=1)\n",
    "netphospan_test['pos'] += 10\n",
    "netphospan_test = netphospan_test[netphospan_test['id'].isin(features_ids)]  # wtf\n",
    "assert len(netphospan_test) == len(netphospan_test[netphospan_test['id'].isin(features_ids)]), 'Mismatch between features and test set'\n",
    "netphospan_test['y_true'] = netphospan_test.apply(get_truth, axis=1)\n",
    "assert len(netphospan_test[netphospan_test['y_true'].isna()]) == 0\n",
    "print(f'AUC: {roc_auc_score(netphospan_test[\"y_true\"], netphospan_test[\"score\"]):.3f}')\n",
    "netphospan_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>auc</th>\n",
       "      <th>auprc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MusiteDeep</th>\n",
       "      <td>0.758658</td>\n",
       "      <td>0.195745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NetPhos 3.1</th>\n",
       "      <td>0.625522</td>\n",
       "      <td>0.086267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NetPhos 4.0</th>\n",
       "      <td>0.864600</td>\n",
       "      <td>0.303167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NetPhospan</th>\n",
       "      <td>0.739899</td>\n",
       "      <td>0.140581</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  auc     auprc\n",
       "MusiteDeep   0.758658  0.195745\n",
       "NetPhos 3.1  0.625522  0.086267\n",
       "NetPhos 4.0  0.864600  0.303167\n",
       "NetPhospan   0.739899  0.140581"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictors = {\n",
    "    'NetPhos 4.0': netphos4_test,\n",
    "    'MusiteDeep': musitedeep_test,\n",
    "    'NetPhospan': netphospan_test,\n",
    "    'NetPhos 3.1': netphos3_test,\n",
    "}\n",
    "predictor_aucs = {}\n",
    "predictor_auprcs = {}\n",
    "for predictor, df in predictors.items():\n",
    "    # ROC and AUC\n",
    "    auc = roc_auc_score(df[\"y_true\"], df[\"score\"])\n",
    "    fpr, tpr, thresholds = roc_curve(df[\"y_true\"], df[\"score\"])\n",
    "    roc_curve_data = pd.DataFrame({'FPR': fpr, 'TPR': tpr, 'threshold': thresholds})\n",
    "    roc_curve_data.to_csv(f'roc_curves/ROC_curve_{predictor.lower().replace(\" \", \"_\")}_global.csv', sep=',', header=True, index=False)\n",
    "    predictor_aucs[predictor] = auc\n",
    "    \n",
    "    # PRC and AUPRC\n",
    "    auprc = average_precision_score(df[\"y_true\"], df[\"score\"])\n",
    "    precision, recall, thresholds = precision_recall_curve(df[\"y_true\"], df[\"score\"])\n",
    "    precision, recall = precision[:-1], recall[:-1]\n",
    "    prc_curve_data = pd.DataFrame({'precision': precision, 'recall': recall, 'threshold': thresholds})\n",
    "    prc_curve_data.to_csv(f'roc_curves/PRC_curve_{predictor.lower().replace(\" \", \"_\")}_global.csv', sep=',', header=True, index=False)\n",
    "    predictor_auprcs[predictor] = auprc\n",
    "\n",
    "aucs = pd.DataFrame({'auc': predictor_aucs})\n",
    "aucs.to_csv('roc_curves/AUCs_global.csv', header=True, index=True)\n",
    "\n",
    "auprcs = pd.DataFrame({'auprc': predictor_auprcs})\n",
    "auprcs.to_csv('roc_curves/AUPRCs_global.csv', header=True, index=True)\n",
    "pd.merge(aucs, auprcs, left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Make inner joins for organism group**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Went from 205415 phosphos to 189180 phosphos\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.472</td>\n",
       "      <td>0</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>5</td>\n",
       "      <td>0.551</td>\n",
       "      <td>0</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>9</td>\n",
       "      <td>0.474</td>\n",
       "      <td>0</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>17</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q5PPN7</td>\n",
       "      <td>24</td>\n",
       "      <td>0.395</td>\n",
       "      <td>0</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189175</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>195</td>\n",
       "      <td>0.232</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189176</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>196</td>\n",
       "      <td>0.247</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189177</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>204</td>\n",
       "      <td>0.128</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189178</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>206</td>\n",
       "      <td>0.105</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189179</th>\n",
       "      <td>Q58FZ9</td>\n",
       "      <td>209</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>189180 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id  pos  score  y_true      group\n",
       "0       Q5PPN7    2  0.472       0  mammalian\n",
       "1       Q5PPN7    5  0.551       0  mammalian\n",
       "2       Q5PPN7    9  0.474       0  mammalian\n",
       "3       Q5PPN7   17  0.433       0  mammalian\n",
       "4       Q5PPN7   24  0.395       0  mammalian\n",
       "...        ...  ...    ...     ...        ...\n",
       "189175  Q58FZ9  195  0.232       0      plant\n",
       "189176  Q58FZ9  196  0.247       0      plant\n",
       "189177  Q58FZ9  204  0.128       0      plant\n",
       "189178  Q58FZ9  206  0.105       0      plant\n",
       "189179  Q58FZ9  209  0.079       0      plant\n",
       "\n",
       "[189180 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "netphos4_test_with_group = netphos4_test.merge(seq_organisms.drop(['phylogeny', 'organism'], axis=1), on='id', how='inner')\n",
    "print(f'Went from {len(netphos4_test)} phosphos to {len(netphos4_test_with_group)} phosphos')\n",
    "netphos4_test_with_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Went from 98461 phosphos to 91830 phosphos\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P73120</td>\n",
       "      <td>12</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P73120</td>\n",
       "      <td>14</td>\n",
       "      <td>0.530</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P73120</td>\n",
       "      <td>20</td>\n",
       "      <td>0.264</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P73120</td>\n",
       "      <td>25</td>\n",
       "      <td>0.138</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P73120</td>\n",
       "      <td>28</td>\n",
       "      <td>0.194</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91825</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>353</td>\n",
       "      <td>0.503</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91826</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>355</td>\n",
       "      <td>0.654</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91827</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>377</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91828</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>387</td>\n",
       "      <td>0.667</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91829</th>\n",
       "      <td>P0DPA2</td>\n",
       "      <td>395</td>\n",
       "      <td>0.704</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91830 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  pos  score  y_true     group\n",
       "0      P73120   12  0.240       0  bacteria\n",
       "1      P73120   14  0.530       0  bacteria\n",
       "2      P73120   20  0.264       0  bacteria\n",
       "3      P73120   25  0.138       0  bacteria\n",
       "4      P73120   28  0.194       0  bacteria\n",
       "...       ...  ...    ...     ...       ...\n",
       "91825  P0DPA2  353  0.503       0     human\n",
       "91826  P0DPA2  355  0.654       0     human\n",
       "91827  P0DPA2  377  0.064       0     human\n",
       "91828  P0DPA2  387  0.667       1     human\n",
       "91829  P0DPA2  395  0.704       0     human\n",
       "\n",
       "[91830 rows x 5 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "musitedeep_test_with_group = musitedeep_test.merge(seq_organisms.drop(['phylogeny', 'organism'], axis=1), on='id', how='inner')\n",
    "print(f'Went from {len(musitedeep_test)} phosphos to {len(musitedeep_test_with_group)} phosphos')\n",
    "musitedeep_test_with_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Went from 35459 phosphos to 30661 phosphos\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0H3MAG4</td>\n",
       "      <td>45</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0H3MAG4</td>\n",
       "      <td>49</td>\n",
       "      <td>0.102</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0H3MAG4</td>\n",
       "      <td>53</td>\n",
       "      <td>0.117</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0H3MAG4</td>\n",
       "      <td>57</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0H3MAG4</td>\n",
       "      <td>63</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30656</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>464</td>\n",
       "      <td>0.224</td>\n",
       "      <td>0</td>\n",
       "      <td>other_animal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30657</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>465</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0</td>\n",
       "      <td>other_animal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30658</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>467</td>\n",
       "      <td>0.172</td>\n",
       "      <td>0</td>\n",
       "      <td>other_animal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30659</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>469</td>\n",
       "      <td>0.097</td>\n",
       "      <td>0</td>\n",
       "      <td>other_animal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30660</th>\n",
       "      <td>Q9GU50</td>\n",
       "      <td>475</td>\n",
       "      <td>0.137</td>\n",
       "      <td>0</td>\n",
       "      <td>other_animal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30661 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  pos  score  y_true         group\n",
       "0      A0A0H3MAG4   45  0.063       0      bacteria\n",
       "1      A0A0H3MAG4   49  0.102       0      bacteria\n",
       "2      A0A0H3MAG4   53  0.117       0      bacteria\n",
       "3      A0A0H3MAG4   57  0.162       0      bacteria\n",
       "4      A0A0H3MAG4   63  0.095       0      bacteria\n",
       "...           ...  ...    ...     ...           ...\n",
       "30656      Q9GU50  464  0.224       0  other_animal\n",
       "30657      Q9GU50  465  0.203       0  other_animal\n",
       "30658      Q9GU50  467  0.172       0  other_animal\n",
       "30659      Q9GU50  469  0.097       0  other_animal\n",
       "30660      Q9GU50  475  0.137       0  other_animal\n",
       "\n",
       "[30661 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "netphospan_test_with_group = netphospan_test.merge(seq_organisms.drop(['phylogeny', 'organism'], axis=1), on='id', how='inner')\n",
    "print(f'Went from {len(netphospan_test)} phosphos to {len(netphospan_test_with_group)} phosphos')\n",
    "netphospan_test_with_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Went from 107469 phosphos to 99678 phosphos\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>score</th>\n",
       "      <th>y_true</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A8IQA2</td>\n",
       "      <td>5</td>\n",
       "      <td>0.399</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A8IQA2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.661</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A8IQA2</td>\n",
       "      <td>11</td>\n",
       "      <td>0.918</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A8IQA2</td>\n",
       "      <td>22</td>\n",
       "      <td>0.449</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A8IQA2</td>\n",
       "      <td>25</td>\n",
       "      <td>0.668</td>\n",
       "      <td>0</td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99673</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2732</td>\n",
       "      <td>0.816</td>\n",
       "      <td>0</td>\n",
       "      <td>other_eukaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99674</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2743</td>\n",
       "      <td>0.660</td>\n",
       "      <td>0</td>\n",
       "      <td>other_eukaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99675</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2753</td>\n",
       "      <td>0.544</td>\n",
       "      <td>0</td>\n",
       "      <td>other_eukaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99676</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2759</td>\n",
       "      <td>0.439</td>\n",
       "      <td>0</td>\n",
       "      <td>other_eukaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99677</th>\n",
       "      <td>V4Z445</td>\n",
       "      <td>2772</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0</td>\n",
       "      <td>other_eukaryotes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99678 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id   pos  score  y_true             group\n",
       "0      A8IQA2     5  0.399       0             plant\n",
       "1      A8IQA2     6  0.661       0             plant\n",
       "2      A8IQA2    11  0.918       0             plant\n",
       "3      A8IQA2    22  0.449       0             plant\n",
       "4      A8IQA2    25  0.668       0             plant\n",
       "...       ...   ...    ...     ...               ...\n",
       "99673  V4Z445  2732  0.816       0  other_eukaryotes\n",
       "99674  V4Z445  2743  0.660       0  other_eukaryotes\n",
       "99675  V4Z445  2753  0.544       0  other_eukaryotes\n",
       "99676  V4Z445  2759  0.439       0  other_eukaryotes\n",
       "99677  V4Z445  2772  0.403       0  other_eukaryotes\n",
       "\n",
       "[99678 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "netphos3_test_with_group = netphos3_test.merge(seq_organisms.drop(['phylogeny', 'organism'], axis=1), on='id', how='inner')\n",
    "print(f'Went from {len(netphos3_test)} phosphos to {len(netphos3_test_with_group)} phosphos')\n",
    "netphos3_test_with_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human</th>\n",
       "      <th>mammalian</th>\n",
       "      <th>mammalian_w_human</th>\n",
       "      <th>other_animal</th>\n",
       "      <th>bacteria</th>\n",
       "      <th>fungal</th>\n",
       "      <th>plant</th>\n",
       "      <th>other_eukaryotes</th>\n",
       "      <th>virus</th>\n",
       "      <th>archaea</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NetPhos 4.0</th>\n",
       "      <td>0.75</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MusiteDeep</th>\n",
       "      <td>0.68</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NetPhospan</th>\n",
       "      <td>0.67</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NetPhos 3.1</th>\n",
       "      <td>0.61</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.51</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             human  mammalian  mammalian_w_human  other_animal  bacteria  \\\n",
       "NetPhos 4.0   0.75       0.81               0.77          0.88      0.84   \n",
       "MusiteDeep    0.68       0.78               0.72          0.82      0.65   \n",
       "NetPhospan    0.67       0.77               0.73          0.73      0.71   \n",
       "NetPhos 3.1   0.61       0.62               0.61          0.66      0.60   \n",
       "\n",
       "             fungal  plant  other_eukaryotes  virus  archaea  \n",
       "NetPhos 4.0    0.86   0.83              0.93   0.73     0.80  \n",
       "MusiteDeep     0.79   0.78              0.78   0.95     0.82  \n",
       "NetPhospan     0.77   0.74              0.77   0.95     0.83  \n",
       "NetPhos 3.1    0.64   0.64              0.61   0.51     0.51  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groups = ['human', 'mammalian', 'mammalian_w_human', 'other_animal', 'bacteria', 'fungal', 'plant', 'other_eukaryotes', 'virus', 'archaea']\n",
    "predictors = {\n",
    "    'NetPhos 4.0': netphos4_test_with_group,\n",
    "    'MusiteDeep': musitedeep_test_with_group,\n",
    "    'NetPhospan': netphospan_test_with_group,\n",
    "    'NetPhos 3.1': netphos3_test_with_group,\n",
    "}\n",
    "predictor_group_aucs = {}\n",
    "for predictor, df in predictors.items():\n",
    "    group_aucs = {}\n",
    "    for group in groups:\n",
    "        if group != 'mammalian_w_human':\n",
    "            group_subset = df[df['group'] == group]\n",
    "        else:\n",
    "            group_subset = df[df['group'].isin(('human', 'mammalian'))]\n",
    "            \n",
    "        if len(group_subset) > 0:\n",
    "            auc = roc_auc_score(group_subset[\"y_true\"], group_subset[\"score\"])\n",
    "            fpr, tpr, thresholds = roc_curve(group_subset[\"y_true\"], group_subset[\"score\"])\n",
    "            roc_curve_data = pd.DataFrame({'FPR': fpr, 'TPR': tpr, 'threshold': thresholds})\n",
    "            roc_curve_data.to_csv(f'roc_curves/ROC_curve_{predictor.lower().replace(\" \", \"_\")}_{group}.csv', header=True, index=False)\n",
    "        else:\n",
    "            auc = 0\n",
    "        group_aucs[group] = round(auc, 2)\n",
    "    predictor_group_aucs[predictor] = group_aucs\n",
    "\n",
    "aucs = pd.DataFrame(predictor_group_aucs).transpose()\n",
    "aucs.to_csv('roc_curves/AUCs_group.csv', header=True, index=True)\n",
    "aucs\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "torch_pyenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
